Data/IntMap/Lazy.hs

```diff
@@ -206,18 +184,1302 @@ module Data.IntMap.Lazy (
     -- * Debugging
     , showTree
     , showTreeWith
-    ) where
+    , valid
+) where
+
+import Data.IntMap.Internal
+import qualified Data.IntMap.Merge.Lazy as Merge (merge, mapMaybeMissing, zipWithMaybeMatched)
+
+import Control.Applicative (Applicative(..))
+import Data.Functor ((<$>))
+
+import Utils.Containers.Internal.StrictPair (StrictPair(..), toPair)
+
+import qualified Data.List (foldl', map)
+import qualified Data.IntSet (IntSet, toList)
+
+import Prelude hiding (foldr, foldl, lookup, null, map, filter, min, max)
+
+-- | /O(1)/. A map of one element.
+--
+-- > singleton 1 'a'        == fromList [(1, 'a')]
+-- > size (singleton 1 'a') == 1
+singleton :: Key -> a -> IntMap a
+singleton k v = IntMap (NonEmpty k v Tip)
+
+-- | /O(min(n,W))/. Insert a new key\/value pair in the map.
+-- If the key is already present in the map, the associated value is
+-- replaced with the supplied value, i.e. 'insert' is equivalent to
+-- @'insertWith' 'const'@.
+--
+-- > insert 5 'x' (fromList [(5,'a'), (3,'b')]) == fromList [(3, 'b'), (5, 'x')]
+-- > insert 7 'x' (fromList [(5,'a'), (3,'b')]) == fromList [(3, 'b'), (5, 'a'), (7, 'x')]
+-- > insert 5 'x' empty                         == singleton 5 'x'
+insert :: Key -> a -> IntMap a -> IntMap a
+insert = start
```

Comment:
Binding them early causes a significant performance penalty, causing a slowdown relative the the current `Data.IntMap` - this was actually one of the big sticking points that kept me from submitting this work as a PR. See [this commit](https://github.com/gereeter/bounded-intmap/commit/8e63ba15041924f53f8eb64c6adf6406226dd567).